### ISSUE TITLE
Docker single executable / memory consumption 

### ISSUE THREAD

**Issue by predaytor** [denoland/fresh#2757]
Is it possible to reduce memory consumption with Fresh 2.0? Currently deployed on fly.io on a 1024MB machine, while a similar application on Node.js consumes around 190-220MB of RAM. Maybe we can create a single executable and get rid of typescript transpilation at runtime? Very thanks.
```Dockerfile
FROM denoland/deno:debian

WORKDIR /app

COPY . .

RUN deno task build

EXPOSE 80

CMD ["run", "-A", "main.ts"]
```

**Comment by predaytor** [comment]
Will this be possible with the release of Deno 2.1.0?

https://github.com/denoland/deno/pull/26939

> To include files or directories in the executable, specify them via the `--include <path>` flag.
> 
> ```shell
> deno compile --include names.csv --include data main.ts
> ```
> 
> Then read the file relative to the directory path of the current module via `import.meta.dirname`:
> 
> ```ts
> // main.ts
> const names = Deno.readTextFileSync(import.meta.dirname + "/names.csv");
> const dataFiles = Deno.readDirSync(import.meta.dirname + "/data");
> 
> // use names and dataFiles here
> ```
> 
> Note this currently only works for files on the file system and not remote
> files.

**Comment by janjangao** [comment]
@predaytor hi, I think since Deno 2.1.3, `--include <path>` shall work correctly with `deno compile`

You can compile the fresh app with single executable file then copy and run in a low memory environment. Can take reference of my Dockerfile
https://github.com/janjangao/hayond.site/blob/main/Dockerfile

**Comment by janjangao** [comment]
> You can compile the fresh app with single executable file then copy and run in a low memory environment.

But there is an other issue, I can see a simple fresh app with 50M at initial state, but the memory increase little by little when day goes by, maybe there's potential memory leak problem.

<img width="579" alt="image" src="https://github.com/user-attachments/assets/4f6332c5-cfba-40b3-ab3e-7520ec436f19" />

**Comment by predaytor** [comment]
@janjangao interesting. for some reason I am getting "Not Found" for my routes. The directly described API endpoints (GET/POST) are working properly:

*upd*: it's a canary version

```tsx
import { App, fsRoutes, staticFiles, trailingSlashes } from 'fresh';
import { satoriApi } from '~/api/satori.tsx';
import { subscribeApi } from '~/api/subscribe.ts';
import { cors } from '~/middleware/cors.ts';
import { secureHeaders } from '~/middleware/security-headers.ts';

export const app = new App<State>({ root: import.meta.url });

app.use(staticFiles());
app.use(trailingSlashes('never'));
app.use(cors);
app.use(secureHeaders);

app.post('/api/subscribe', subscribeApi);
app.get('/api/satori', satoriApi);

await fsRoutes(app, {
    dir: './',
    loadIsland: (path) => import(`./islands/${path}`),
    loadRoute: (path) => import(`./routes/${path}`),
});

if (import.meta.main) {
    await app.listen();
}
```

**Comment by predaytor** [comment]
on Canary it's about ~370 MB of memory consumption, in addition to static files and routes not working.

**Comment by janjangao** [comment]
> for some reason I am getting "Not Found" for my routes.

oh, I didn't write complex logic like this, maybe let me proceed try my webside to see if have same issue.

**Comment by janjangao** [comment]
> upd: it's a canary version

also, let me try the canary version later

**Comment by predaytor** [comment]
With the new Deno 2.3.1 version, memory consumption has been significantly reduced compared. The compiled version also works properly, but I was not able to deploy it to fly.io to check the exact difference in memory usage.
```Dockerfile
FROM denoland/deno:debian AS builder
WORKDIR /app

COPY ./web ./
RUN deno task build
RUN deno compile --env=./.env --include _fresh/ --include deno.json -A -o exe main.ts

# FROM gcr.io/distroless/cc
# FROM denoland/deno:distroless
FROM denoland/deno:debian
COPY --from=builder /app/exe ./
EXPOSE 8000
CMD ["./exe"]
```

fly logs:
```bash
Waiting for logs...

2025-05-09T18:33:20.507 proxy[286ee3efe03de8] ams [info] Starting machine

2025-05-09T18:33:20.618 app[286ee3efe03de8] ams [info] 2025-05-09T18:33:20.618502415 [01JTV4Y34D66E79DSVX0447E3Z:main] Running Firecracker v1.7.0

2025-05-09T18:33:21.557 app[286ee3efe03de8] ams [info] INFO Starting init (commit: a1a45272)...

2025-05-09T18:33:21.695 app[286ee3efe03de8] ams [info] INFO Preparing to run: `/tini -- docker-entrypoint.sh ./exe` as root

2025-05-09T18:33:21.703 app[286ee3efe03de8] ams [info] INFO [fly api proxy] listening at /.fly/api

2025-05-09T18:33:21.708 runner[286ee3efe03de8] ams [info] Machine started in 1.196s

2025-05-09T18:33:21.709 proxy[286ee3efe03de8] ams [info] machine started in 1.202416794s

2025-05-09T18:33:21.714 app[286ee3efe03de8] ams [info] [WARN tini (628)] Tini is not running as PID 1 and isn't registered as a child subreaper.

2025-05-09T18:33:21.714 app[286ee3efe03de8] ams [info] Zombie processes will not be re-parented to Tini, so zombie reaping won't work.

2025-05-09T18:33:21.714 app[286ee3efe03de8] ams [info] To fix the problem, use the -s option or set the environment variable TINI_SUBREAPER to register Tini as a child subreaper, or run Tini as PID 1.

2025-05-09T18:33:22.197 app[286ee3efe03de8] ams [info] 2025/05/09 18:33:22 INFO SSH listening listen_address=[fdaa:0:ebf7:a7b:3f7:a10e:653b:2]:22

2025-05-09T18:33:27.167 proxy[286ee3efe03de8] ams [info] waiting for machine to be reachable on 0.0.0.0:8000 (waited 5.457763703s so far)

2025-05-09T18:33:30.195 proxy[286ee3efe03de8] ams [error] [PR03] could not find a good candidate within 1 attempts at load balancing. last error: [PM05] failed to connect to machine: gave up after 15 attempts (in 8.465035978s)

2025-05-09T18:33:30.302 proxy[286ee3efe03de8] ams [error] [PC01] instance refused connection. is your app listening on 0.0.0.0:8000? make sure it is not only listening on 127.0.0.1 (hint: look at your startup logs, servers often print the address they are listening on)

2025-05-09T18:33:35.122 app[286ee3efe03de8] ams [info] [ 14.405715] Out of memory: Killed process 632 (exe) total-vm:555508kB, anon-rss:400768kB, file-rss:128kB, shmem-rss:0kB, UID:0 pgtables:884kB oom_score_adj:0

2025-05-09T18:33:35.717 app[286ee3efe03de8] ams [info] INFO Main child exited normally with code: 137

2025-05-09T18:33:35.734 app[286ee3efe03de8] ams [info] INFO Starting clean up.

2025-05-09T18:33:35.735 app[286ee3efe03de8] ams [info] WARN could not unmount /rootfs: EINVAL: Invalid argument

2025-05-09T18:33:35.736 app[286ee3efe03de8] ams [info] [ 15.021069] reboot: Restarting system

2025-05-09T18:33:37.592 app[286ee3efe03de8] ams [info] 2025-05-09T18:33:37.592593280 [01JTV4Y34D66E79DSVX0447E3Z:main] Running Firecracker v1.7.0
```

**Comment by janjangao** [comment]
with my checking, seems the Deno 2.3.1 has slight(not so much) improvement, the footprint size reduced from 50m to 46.8m, see [tag comparison 0.0.2 with 0.0.1](https://hub.docker.com/repository/docker/janjangao/hayond.site/tags), the memory deduction is 55m by 61m. 

anyway, maybe my app is super small, there's no any space to reduce, but it's worth to update to new version.

> FROM denoland/deno:debian

I noticed you are using deno:debian as base image, I think it's not necessary, since after compilation, deno runtime is no use anymore,  [distroless/cc](https://github.com/GoogleContainerTools/distroless/tree/main/cc) shall has the minimal native runtime like `libgcc`

**Comment by predaytor** [comment]
> I noticed you are using deno:debian as base image, I think it's not necessary, since after compilation, deno runtime is no use anymore, [distroless/cc](https://github.com/GoogleContainerTools/distroless/tree/main/cc) shall has the minimal native runtime like `libgcc`

yes, the point is that when using multi-stage build (copying only the executable to the final image), I got `Not Found` for all routes in the `/routes` file router. I think it is still trying to reference files from the *outside* world, other than the embedded files in the .exe itself.

**Comment by predaytor** [comment]
> I noticed you are using deno:debian as base image, I think it's not necessary, since after compilation, deno runtime is no use anymore, [distroless/cc](https://github.com/GoogleContainerTools/distroless/tree/main/cc) shall has the minimal native runtime like `libgcc`

sorry it seems when using multi-stage build (copying only the executable to the final image), I got `Not Found` for all routes in the `/routes` file router. I think it is still trying to reference files from the *outside* world, other than the embedded files in the .exe itself.

**Comment by hronro** [comment]
@predaytor Exactly. I can reproduce this without Docker. Simply move the executable to somewhere else then run the executable:

```
deno compile --include _fresh/ --include deno.json -A -o exe main.ts
mv exe /tmp/
cd /tmp/
./exe
```

I create a new issue for this: https://github.com/denoland/fresh/issues/3123

**Comment by bartlomieju** [comment]
Closing this — the main issues raised here have been addressed:

- **`deno compile` support**: Fresh now works with `deno compile` (the routes-missing issue was fixed in #3123)
- **Memory consumption**: Has improved with newer Deno versions (2.3.1+). The remaining memory overhead is Deno runtime behavior, not Fresh-specific
- **Docker setup**: The thread has working examples for both `deno compile` and `deno serve` based Docker deployments

The potential memory leak mentioned by @janjangao would be better tracked as a separate issue with a minimal reproduction if it's still occurring on recent Deno versions.
